### PR TITLE
1445 Patch

### LINKED ISSUE #1445: EXC_BAD_ACCESS (code=EXC_I386_GPFLT) in trick_error_report() when terminating during init on Darwin
Discovered as part of my work on #1415, I can replicate the following segmentation fault on my Darwin build (Mac OSX 12.6.2) with [config.log](https://github.com/nasa/trick/files/10550032/config.log) and `S_define`:

```c++
#include "sim_objects/default_trick_sys.sm"
##include "trick/exec_proto.h"

class EarlyTerminationSimObject : public Trick::SimObject
{
 public:
  double x;

  EarlyTerminationSimObject()
    :
    x(0)
  {
    ("initialization") early_term();
  };
  void early_term() {
    std::string message = "Terminating with exit code 0";
    exec_terminate_with_return(0, "S_define", 24, message.c_str());
  }
 private:
  EarlyTerminationSimObject( const EarlyTerminationSimObject&);
  EarlyTerminationSimObject & operator= ( const EarlyTerminationSimObject&);
};
EarlyTerminationSimObject test;
```
... and an empty `RUN_test/input.py`. This results in:

```
me@host: ~/dev/trick/test/SIM_earlyterm > ./S_main_Darwin_21.exe RUN_test/input.py 
|L   0|2023/01/31,13:18:54|JSLAL0520093471| |T 0|0.000000| 
SIMULATION TERMINATED IN
  PROCESS: 0
  ROUTINE: S_define:24
  DIAGNOSTIC: Terminating with exit code 0

       SIMULATION START TIME:        0.000
        SIMULATION STOP TIME:        0.000
     SIMULATION ELAPSED TIME:        0.000
        ACTUAL CPU TIME USED:        0.240
       SIMULATION / CPU TIME:        0.000
     INITIALIZATION CPU TIME:        0.000
Process terminated by signal SIGSEGV
```
Using `lldb` for the same configuration, I get the following stacktrace:

```
Process 14766 stopped
* thread #3, stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x0000000100555184 S_main_Darwin_21.exe`trick_error_report + 436
S_main_Darwin_21.exe`trick_error_report:
->  0x100555184 <+436>: callq  *%rax
    0x100555186 <+438>: jmp    0x10055518b               ; <+443>
    0x10055518b <+443>: movq   0x21f50e(%rip), %rax      ; (void *)0x00007ff848b2b8a0: __stack_chk_guard
    0x100555192 <+450>: movq   (%rax), %rax
Target 0: (S_main_Darwin_21.exe) stopped.
(lldb) thread backtrace all
  thread #1, queue = 'com.apple.main-thread'
    frame #0: 0x00007ff80739ad96 libsystem_kernel.dylib`__close_nocancel + 10
    frame #1: 0x00007ff8072c59ef libsystem_c.dylib`fclose + 107
    frame #2: 0x00007ff80734aa99 libc++.1.dylib`std::__1::basic_filebuf<char, std::__1::char_traits<char> >::close() + 39
    frame #3: 0x000000010017444d S_main_Darwin_21.exe`std::__1::basic_fstream<char, std::__1::char_traits<char> >::close() + 29
    frame #4: 0x0000000100174377 S_main_Darwin_21.exe`Trick::MessageFile::~MessageFile() + 39
    frame #5: 0x00000001001744b5 S_main_Darwin_21.exe`Trick::MessageFile::~MessageFile() + 21
    frame #6: 0x0000000100012d24 S_main_Darwin_21.exe`MessageSimObject::~MessageSimObject() + 84
    frame #7: 0x0000000100003505 S_main_Darwin_21.exe`MessageSimObject::~MessageSimObject() + 21
    frame #8: 0x00007ff8072cede4 libsystem_c.dylib`__cxa_finalize_ranges + 409
    frame #9: 0x00007ff8072cebfe libsystem_c.dylib`exit + 35
    frame #10: 0x00007ff8073e2375 libdyld.dylib`dyld4::LibSystemHelpers::exit(int) const + 11
    frame #11: 0x0000000100bbd558 dyld`start + 504
  thread #2
    frame #0: 0x00007ff8073a1d5a libsystem_kernel.dylib`__select + 10
    frame #1: 0x00000001001dc6ac S_main_Darwin_21.exe`Trick::VariableServerListenThread::thread_body() + 1100
    frame #2: 0x00000001001ce3d0 S_main_Darwin_21.exe`Trick::ThreadBase::thread_helper(void*) + 96
    frame #3: 0x00007ff8073d64e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #4: 0x00007ff8073d1f6b libsystem_pthread.dylib`thread_start + 15
* thread #3, stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x0000000100555184 S_main_Darwin_21.exe`trick_error_report + 436
    frame #1: 0x000000010054f335 S_main_Darwin_21.exe`tc_accept_ + 245
    frame #2: 0x000000010017ae36 S_main_Darwin_21.exe`Trick::MessageTCDeviceListenThread::thread_body() + 694
    frame #3: 0x00000001001ce3d0 S_main_Darwin_21.exe`Trick::ThreadBase::thread_helper(void*) + 96
    frame #4: 0x00007ff8073d64e1 libsystem_pthread.dylib`_pthread_start + 125
    frame #5: 0x00007ff8073d1f6b libsystem_pthread.dylib`thread_start + 15
```
This setup is pushed to `origin/earlyterm_darwin_issue` and is currently blocking successful Darwin testing of the MCG work in #1415, which is a model that terminates during init for nominal use cases. I am unable to replicate this on CentOS 7.10 and no other platform in GitHub actions seems to detect it either.

FYI @jdeans289 @jmpenn

### PR DESCRIPTION
**Pull request by jdeans289**
Closes #1445 

Resolves a race condition between the destructor of `MessageTCDeviceListenThread` and the thread itself. When a sim is terminated during initialization, the shutdown jobs do not run. As a result, some threads, like the MessageTCDeviceListenThread, may have been spun up but not cancelled. When the data they rely on is destroyed, lots of different segfaults can occur. This pull request attempts to solve that problem by making sure the threads are finished at the beginning of the destructor.

### CODE REVIEW

**Comment by jdeans289** [comment]
Something odd is happening with the Mac pipeline, and I can't tell if it's something I did or the newest upgrade to the runner. I suspect the runner.

File: include/trick/SysThread.hh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * The purpose of this class is to ensure safe shutdown for Trick system threads, since user threads are handled separately in
+     * the Trick::Threads and Executive classes.
+     *
+     * @author Jackie Deans
+     *
+     *      
+     **/
+    class SysThread : public Trick::ThreadBase {
```
Comment (jmpenn) [inline include/trick/SysThread.hh:32]: Could you add a comment that references issue 1445, that specifies the motivation for creating this class?

**Review by jmpenn — COMMENTED** [review]
Just the comment for SysThread class that refers to issue 1445

**Comment by Codym48** [comment]
Heads up, @jdeans289, we think this might be causing #1656

**Pull request merged**